spec/formatting.md

```diff
@@ -23,11 +23,15 @@ are treated as identical.
 To resolve the value of a _variable_,
 its _name_ is used to identify either a local variable,
 or a variable defined elsewhere.
-If a local variable and an externally defined one use the same name,
-the local variable takes precedence.
 
-It is an error for a local variable definition to
-refer to a local variable that's defined after it in the message.
+It is an error for the right-hand side of a local variable declaration to
+refer to a local variable that's declared after it in the message.
+
+Variable names are required to be globally unique. That is,
+for any local variable declaration of the form `let $v = e`:
+* It is an error if `v` is the left-hand side of a local variable declaration
+that appears before `let $v = e` in the message.
+* It is also an error if `v` is an externally defined variable.
```

Comment:
We discussed this briefly in the recent call (talking about changing the sigil for external values). I think this is bad because it is easy for code to suddenly produce an error when the input data (which is dynamic) collides with a declaration.

Reply:
Nitpick: `let $v = e` => `let $v = {$e}` (everywhere)

Reply:
@aphillips I agree that it would be preferable to change the sigil for externally-defined variables. If that change lands before this PR does (i.e. if #403 is resolved), then I'll update this PR to reflect that.

@mihnita I think my rewrites that I just pushed eliminate metavariables anyway, but if I end up re-adding any I'll keep that in mind.